README.md
# json-response

[![Build Status](https://travis-ci.com/CaliforniaMountainSnake/json-response.svg?branch=master)](https://travis-ci.com/CaliforniaMountainSnake/json-response)

A simple object for creating api responses.

## Install:
### Require this package with Composer
Install this package through [Composer](https://getcomposer.org/).
Edit your project's `composer.json` file to require `californiamountainsnake/json-response`:
```json
{
    "name": "yourproject/yourproject",
    "type": "project",
    "require": {
        "php": "^7.1",
        "californiamountainsnake/json-response": "*"
    }
}
```
and run `composer update`

### or
run this command in your command line:
```bash
composer require californiamountainsnake/json-response
```

## Usage examples:
```php
<?php
use CaliforniaMountainSnake\JsonResponse\JsonResponse;
return JsonResponse::error(['error_1', 'error_2', 'error_3'], JsonResponse::HTTP_BAD_REQUEST)->make();
```


CHANGELOG.md
# Changelog
The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security


## [1.0.5] - 2020-05-23
### Fixed
- Fixed "illuminate/http" dependency.
- composer.lock has been deleted from the repository.

## [1.0.4] - 2020-05-23
### Fixed
- Fixed "illuminate/http" dependency.

## [1.0.3] - 2020-05-23
### Added
- Added Travis-CI integration.
### Fixed
- Fixed "illuminate/http" dependency.

## [1.0.2] - 2019-09-05
### Added
- The header "Content-Type: application/json" has been permanently added to the each response.

## [1.0.1] - 2019-07-31
### Added
- Added the possibility to add custom http headers to the response.
- Added the HttpCodes interface.
- Added getters.
- Added this log.
### Changed
- Minimal php version was reduced to 7.1.
- Updated the Composer dependencies.

[1.0.5]: https://github.com/CaliforniaMountainSnake/json-response/compare/1.0.4...1.0.5
[1.0.4]: https://github.com/CaliforniaMountainSnake/json-response/compare/1.0.3...1.0.4
[1.0.3]: https://github.com/CaliforniaMountainSnake/json-response/compare/1.0.2...1.0.3
[1.0.2]: https://github.com/CaliforniaMountainSnake/json-response/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/CaliforniaMountainSnake/json-response/compare/fc1c036571c27213de85d7c21a447144bc4947b1...1.0.1


tests/Unit/JsonResponseTest.php
<?php

namespace Tests\Unit;

use CaliforniaMountainSnake\JsonResponse\JsonResponse;
use PHPUnit\Framework\TestCase;

class JsonResponseTest extends TestCase
{
    /**
     * @throws \SebastianBergmann\RecursionContext\InvalidArgumentException
     */
    public function testToJson(): void
    {
        $info = [
            'key1' => 'value1',
            'key2' => 'value2',
        ];

        $json = JsonResponse::good($info)->toJson();
        $this->assertEquals('{"is_ok":"true","body":{"key1":"value1","key2":"value2"}}', $json);
    }

    /**
     * @throws \SebastianBergmann\RecursionContext\InvalidArgumentException
     */
    public function testDefaultNoHeadersRequest(): void
    {
        $jr = JsonResponse::good(['test']);

        // Content-Type: application/json is in the every response.
        $this->assertSame(\count($jr->getHeaders()), 1);
    }

    /**
     * @throws \SebastianBergmann\RecursionContext\InvalidArgumentException
     */
    public function testHeaders(): void
    {
        $jr = JsonResponse::good(['test'])->withCors()->headers(['custom_name' => 'custom_value']);
        $this->assertSame(\count($jr->getHeaders()), 5);
    }
}


src/HttpCodes.php
<?php

namespace CaliforniaMountainSnake\JsonResponse;

/**
 * Http response codes.
 * (From Symfony\Component\HttpFoundation\Response).
 */
interface HttpCodes
{
    public const HTTP_CONTINUE                             = 100;
    public const HTTP_SWITCHING_PROTOCOLS                  = 101;
    public const HTTP_PROCESSING                           = 102; // RFC2518
    public const HTTP_EARLY_HINTS                          = 103; // RFC8297
    public const HTTP_OK                                   = 200;
    public const HTTP_CREATED                              = 201;
    public const HTTP_ACCEPTED                             = 202;
    public const HTTP_NON_AUTHORITATIVE_INFORMATION        = 203;
    public const HTTP_NO_CONTENT                           = 204;
    public const HTTP_RESET_CONTENT                        = 205;
    public const HTTP_PARTIAL_CONTENT                      = 206;
    public const HTTP_MULTI_STATUS                         = 207; // RFC4918
    public const HTTP_ALREADY_REPORTED                     = 208; // RFC5842
    public const HTTP_IM_USED                              = 226; // RFC3229
    public const HTTP_MULTIPLE_CHOICES                     = 300;
    public const HTTP_MOVED_PERMANENTLY                    = 301;
    public const HTTP_FOUND                                = 302;
    public const HTTP_SEE_OTHER                            = 303;
    public const HTTP_NOT_MODIFIED                         = 304;
    public const HTTP_USE_PROXY                            = 305;
    public const HTTP_RESERVED                             = 306;
    public const HTTP_TEMPORARY_REDIRECT                   = 307;
    public const HTTP_PERMANENTLY_REDIRECT                 = 308; // RFC7238
    public const HTTP_BAD_REQUEST                          = 400;
    public const HTTP_UNAUTHORIZED                         = 401;
    public const HTTP_PAYMENT_REQUIRED                     = 402;
    public const HTTP_FORBIDDEN                            = 403;
    public const HTTP_NOT_FOUND                            = 404;
    public const HTTP_METHOD_NOT_ALLOWED                   = 405;
    public const HTTP_NOT_ACCEPTABLE                       = 406;
    public const HTTP_PROXY_AUTHENTICATION_REQUIRED        = 407;
    public const HTTP_REQUEST_TIMEOUT                      = 408;
    public const HTTP_CONFLICT                             = 409;
    public const HTTP_GONE                                 = 410;
    public const HTTP_LENGTH_REQUIRED                      = 411;
    public const HTTP_PRECONDITION_FAILED                  = 412;
    public const HTTP_REQUEST_ENTITY_TOO_LARGE             = 413;
    public const HTTP_REQUEST_URI_TOO_LONG                 = 414;
    public const HTTP_UNSUPPORTED_MEDIA_TYPE               = 415;
    public const HTTP_REQUESTED_RANGE_NOT_SATISFIABLE      = 416;
    public const HTTP_EXPECTATION_FAILED                   = 417;
    public const HTTP_I_AM_A_TEAPOT                        = 418; // RFC2324
    public const HTTP_MISDIRECTED_REQUEST                  = 421; // RFC7540
    public const HTTP_UNPROCESSABLE_ENTITY                 = 422; // RFC4918
    public const HTTP_LOCKED                               = 423; // RFC4918
    public const HTTP_FAILED_DEPENDENCY                    = 424; // RFC4918
    public const HTTP_TOO_EARLY                            = 425; // RFC-ietf-httpbis-replay-04
    public const HTTP_UPGRADE_REQUIRED                     = 426; // RFC2817
    public const HTTP_PRECONDITION_REQUIRED                = 428; // RFC6585
    public const HTTP_TOO_MANY_REQUESTS                    = 429; // RFC6585
    public const HTTP_REQUEST_HEADER_FIELDS_TOO_LARGE      = 431; // RFC6585
    public const HTTP_UNAVAILABLE_FOR_LEGAL_REASONS        = 451;
    public const HTTP_INTERNAL_SERVER_ERROR                = 500;
    public const HTTP_NOT_IMPLEMENTED                      = 501;
    public const HTTP_BAD_GATEWAY                          = 502;
    public const HTTP_SERVICE_UNAVAILABLE                  = 503;
    public const HTTP_GATEWAY_TIMEOUT                      = 504;
    public const HTTP_VERSION_NOT_SUPPORTED                = 505;
    public const HTTP_VARIANT_ALSO_NEGOTIATES_EXPERIMENTAL = 506; // RFC2295
    public const HTTP_INSUFFICIENT_STORAGE                 = 507; // RFC4918
    public const HTTP_LOOP_DETECTED                        = 508; // RFC5842
    public const HTTP_NOT_EXTENDED                         = 510; // RFC2774
    public const HTTP_NETWORK_AUTHENTICATION_REQUIRED      = 511; // RFC6585
}


src/JsonResponse.php
<?php

namespace CaliforniaMountainSnake\JsonResponse;

use Illuminate\Http\Response;

/**
 * The json-response object for uniformity of responses api.
 */
class JsonResponse implements HttpCodes
{
    public const IS_OK = 'is_ok';
    public const ERRORS = 'errors';
    public const BODY = 'body';

    /**
     * @var bool
     */
    protected $isOk;

    /**
     * @var int
     */
    protected $httpCode;

    /**
     * @var array
     */
    protected $errors;

    /**
     * @var array|null
     */
    protected $body;

    /**
     * @var string[]
     */
    protected $headers = [];

    /**
     * JsonResponse constructor.
     *
     * @param int        $_http_code
     * @param array      $errors
     * @param array|null $body
     */
    public function __construct(int $_http_code, array $errors = [], array $body = null)
    {
        $this->isOk = true;
        $this->httpCode = $_http_code;
        $this->errors = $errors;
        $this->body = $body;

        if (!empty($errors)) {
            $this->isOk = false;
        }

        $this->withApplicationJson();
    }

    /**
     * Allow cross-origin requests.
     *
     * @return JsonResponse
     */
    public function withCors(): self
    {
        $this->headers([
            'Access-Control-Allow-Origin' => '*',
            'Access-Control-Allow-Methods' => 'GET, POST, PUT, DELETE, OPTIONS',
            'Access-Control-Allow-Headers' => 'X-Requested-With, Content-Type, X-Token-Auth, Authorization',
        ]);

        return $this;
    }

    /**
     * Add custom headers to the response.
     *
     * @param string[] $_headers Array with headers ['header_name' => 'header_value']
     *
     * @return JsonResponse
     */
    public function headers(array $_headers): self
    {
        $this->headers = \array_merge($this->headers, $_headers);
        return $this;
    }

    /**
     * @return Response
     */
    public function make(): Response
    {
        try {
            $response = new Response($this->toJson(), $this->httpCode);
            return $response->withHeaders($this->headers);
        } catch (\LogicException $e) {
            throw new \RuntimeException($e->getMessage(), $e->getCode(), $e);
        }
    }

    /**
     * @param array $body
     * @param int   $_http_code
     *
     * @return JsonResponse
     */
    public static function good(array $body, int $_http_code = 200): self
    {
        return new self($_http_code, [], $body);
    }

    /**
     * @param array $errors
     * @param int   $_http_code
     *
     * @return JsonResponse
     */
    public static function error(array $errors, int $_http_code = 400): self
    {
        return new self ($_http_code, $errors, null);
    }

    /**
     * @return array
     */
    public function toArray(): array
    {
        $arr = [
            self::IS_OK => $this->isOk ? 'true' : 'false',
        ];

        if (!empty($this->errors)) {
            $arr[self::ERRORS] = $this->errors;
        }
        if ($this->body !== null) {
            $arr[self::BODY] = $this->body;
        }

        return $arr;
    }

    /**
     * @return string
     */
    public function toJson(): string
    {
        return \json_encode($this->toArray());
    }

    /**
     * @return array
     */
    public function __debugInfo(): array
    {
        return $this->toArray();
    }

    //------------------------------------------------------------------------------------------------------------------

    /**
     * Add "Content-Type: application/json" header to the response.
     *
     * @return JsonResponse
     */
    private function withApplicationJson(): self
    {
        $this->headers([
            'Content-Type' => 'application/json',
        ]);

        return $this;
    }

    //------------------------------------------------------------------------------------------------------------------
    // Getters.
    //------------------------------------------------------------------------------------------------------------------

    /**
     * @return bool
     */
    public function isOk(): bool
    {
        return $this->isOk;
    }

    /**
     * @return int
     */
    public function getHttpCode(): int
    {
        return $this->httpCode;
    }

    /**
     * @return array
     */
    public function getErrors(): array
    {
        return $this->errors;
    }

    /**
     * @return array|null
     */
    public function getBody(): ?array
    {
        return $this->body;
    }

    /**
     * @return string[]
     */
    public function getHeaders(): array
    {
        return $this->headers;
    }
}
